src/main/java/sparksoniq/jsoniq/runtime/iterator/primary/ObjectConstructorRuntimeIterator.java

```diff
@@ -75,27 +77,32 @@ public ObjectItem next() {
                 for (RuntimeIterator valueIterator : this._values) {
                     List<Item> currentResults = new ArrayList<>();
                     valueIterator.open(this._currentDynamicContext);
-                    while (valueIterator.hasNext())
+                    while (valueIterator.hasNext()) {
                         currentResults.add(valueIterator.next());
-//                    if(valueIterator.hasNext())
-//                        throw new IteratorFlowException("Object value must return one item!");
+                    }
                     valueIterator.close();
                     //SIMILAR TO ZORBA, if value is more than one item, wrap it in an array
-                    if (currentResults.size() > 1)
+                    if (currentResults.size() > 1) {
                         values.add(new ArrayItem(currentResults));
-                    else
+                    } else if (currentResults.size() == 1) {
                         values.add(currentResults.get(0));
+                    } else {
+                        values.add(new NullItem());
+                    }
                 }
 
                 for (RuntimeIterator keyIterator : this._keys) {
                     keyIterator.open(this._currentDynamicContext);
-                    try {
-                        keys.add(((StringItem) keyIterator.next()).getStringValue());
-                    } catch (Exception ex) {
-                        throw new IteratorFlowException("Object must have string keys!", getMetadata());
+                    if (!keyIterator.hasNext()) {
+                        throw new IteratorFlowException("Key iterator must provide an item for object creation", getMetadata());
```

Comment:
Is this message visible outside? If so, it is more data-independent to say "A key cannot be the empty sequence".